api/internal/localizer/localizer.go

```diff
@@ -0,0 +1,364 @@
+	"sigs.k8s.io/kustomize/api/krusty"
+	"sigs.k8s.io/kustomize/api/loader"
+	"sigs.k8s.io/kustomize/api/provider"
+	"sigs.k8s.io/kustomize/api/resmap"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+	"sigs.k8s.io/yaml"
+)
+
+// localizer localizes the kustomization root at ldr.
+type localizer struct {
+	fSys filesys.FileSystem
+
+	ldr ifc.Loader
+
+	// destination directory in newDir that mirrors ldr's current root.
+	dst filesys.ConfirmedDir
+
+	// all localize directories created
+	localizeDirs map[filesys.ConfirmedDir]struct{}
+
+	// kusttarget fields
+	rFactory  *resmap.Factory
+	pLdr      *pluginsLoader.Loader
+	validator ifc.Validator
+}
+
+// Run attempts to localize the kustomization root at targetArg with the given localize arguments
+func Run(fSys filesys.FileSystem, targetArg string, scopeArg string, newDirArg string) error {
+	ldr, args, err := NewLocLoader(targetArg, scopeArg, newDirArg, fSys)
+	if err != nil {
+		return err
+	}
+	defer ldr.Cleanup()
+
+	// create directory that corresponds to target in newDir
+	var realScope string
+	if repo, isRemote := ldr.Repo(); isRemote {
+		realScope = repo
+	} else {
+		realScope = args.Scope.String()
+	}
+	toDst, err := filepath.Rel(realScope, ldr.Root())
+	if err != nil {
+		log.Fatalf(errors.WrapPrefixf(err, errRelativeWhenContains(realScope, ldr.Root())).Error())
+	}
+	dst := args.NewDir.Join(toDst)
+	if err = fSys.MkdirAll(dst); err != nil {
+		return errors.WrapPrefixf(err, "unable to create directory in localize destination")
+	}
+
+	depProvider := provider.NewDepProvider()
+	rFactory := resmap.NewFactory(depProvider.GetResourceFactory())
+	err = (&localizer{
+		fSys:      fSys,
+		ldr:       ldr,
+		dst:       filesys.ConfirmedDir(dst),
+		validator: depProvider.GetFieldValidator(),
+		rFactory:  rFactory,
+		pLdr:      pluginsLoader.NewLoader(krusty.MakeDefaultOptions().PluginConfig, rFactory, filesys.MakeFsOnDisk()),
+	}).localize()
+	if err != nil {
+		if errCleanup := fSys.RemoveAll(args.NewDir.String()); errCleanup != nil {
+			log.Printf("%s", errors.WrapPrefixf(errCleanup, "unable to clean localize destination").Error())
+		}
+		return errors.WrapPrefixf(err, "unable to localize target '%s'", targetArg)
+	}
+	return nil
+}
+
+func (lclzr *localizer) localize() error {
+	kt := target.NewKustTarget(lclzr.ldr, lclzr.validator, lclzr.rFactory, lclzr.pLdr)
+	err := kt.Load()
+	if err != nil {
+		return err
+	}
+
+	kust, err := lclzr.processKust(kt)
+	if err != nil {
+		return err
+	}
+
+	content, err := yaml.Marshal(kust)
+	if err != nil {
+		return errors.WrapPrefixf(
+			err,
+			"unable to serialize localized kustomization file for root '%s'", lclzr.ldr.Root())
+	}
+	return lclzr.fSys.WriteFile(lclzr.dst.Join(konfig.DefaultKustomizationFileName()), content)
+}
+
+func (lclzr *localizer) processKust(kt *target.KustTarget) (*types.Kustomization, error) {
+	kust := kt.Kustomization()
```

Comment:
> adding MapFiles() to each plugin under plugin/builtin

Based on our discussion yesterday, do you plan to try that again? I like this idea the most of the options you've explored but want to understand more: 

> includes re-factoring a lot of the internal code to keep consistent state

Could you provide an example of this? Refactoring should be fine, but depending on what's required, we may have some recommendations. A really large refactor might warrant its own PR, but my impression from our conversation was that it's not so much refactoring that that would be necessary.

> MapFiles() for all the existing plugins,
 I think for the scope of this PR, we would only need to handle the builtin plugins that also have corresponding kustomization fields. So like configmapgenerator, replacements, etc. 

For plugins that we said either we wouldn't handle (like the helm fields), we can have an empty `MapFiles` with a TODO comment.

For plugins that are not associated with kustomization fields, we can tackle in the next PR when you explicitly handle `transformers` and `generators`.

Reply:
> Based on our discussion yesterday, do you plan to try that again? 

Yeah. I thought it was more work than necessary because I was under the false assumption that we don't need to handle go plugins that aren't kustomization fields. Now knowing that we should `localize` builtin plugins under the `generators`, `transformers` fields, I think this design (or something similar) will be necessary. I assume (will check) we currently process `generators`, `transformers` go plugins by finding their corresponding go plugin binary and executing exposed methods, so exposing a `MapFiles()` method on the go plugins should be most scalable.

> but my impression from our conversation was that it's not so much refactoring that that would be necessary.

No, I don't think it should be too bad.

> Could you provide an example of this? 

Sure. I haven't looked through all of them yet, but take [`plugin/builtin/PatchTransformer`](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/patchtransformer/PatchTransformer.go) as an example. It has fields like `loadedPatch`, `decodedPatch` that would need to be re-calculated in `MapFiles` once the internal path is mapped to a different, in our case localized, path. I'd move some of the logic in [`Config`](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/patchtransformer/PatchTransformer.go#L32) out into a private state-checking method that `Config`, `MapFiles` could then call.

There'll probably be some other minor changes as well. For example, I may create a new private field `patch` to hold the loaded patch instead of [modifying the public `Patch`](https://github.com/kubernetes-sigs/kustomize/blob/master/plugin/builtin/patchtransformer/PatchTransformer.go#L52) because modifying public fields will have unintended consequences when we write out the plugin with `yaml.Marshal()`.

Reply:
> A few things to consider with `MapFiles`:
> 
> * I think for the scope of this PR, we would only need to handle the builtin plugins that also have corresponding kustomization fields. So like configmapgenerator, replacements, etc.
> * For plugins that we said we wouldn't handle (like the helm fields), we can have an empty `MapFiles` with a TODO comment.
> * For plugins that are not associated with kustomization fields, we can tackle in the next PR when you explicitly handle `transformers` and `generators`. You could keep these `MapFiles` functions empty with a TODO as well for now.

Thanks for the clarifications and suggestions!